### PR TITLE
Upgrade jackson to 2.9.5

### PR DESCRIPTION
**Pull request by robert3005**

### CODE REVIEW

File: jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/feignimpl/GuavaOptionalAwareDecoderTest.java [hunk 1]
```diff
@@ -132,9 +132,9 @@ public void testComplexType() {
                         new GuavaOptionalComplexType(
                                 Optional.absent(),
                                 Optional.absent(),
-                                Paths.get("bar"))),
+                                Paths.get("/bar"))),
                 Optional.of("baz"),
-                Paths.get("foo"));
+                Paths.get("/foo"));
```
Comment (uschi2000) [inline jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/feignimpl/GuavaOptionalAwareDecoderTest.java]: why this change?

File: jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/feignimpl/Java8TestServer.java [hunk 2]
```diff
@@ -46,6 +47,7 @@
     public final void run(Configuration config, final Environment env) throws Exception {
         env.jersey().register(HttpRemotingJerseyFeature.INSTANCE);
         env.jersey().register(new TestResource());
+        env.jersey().register(new EmptyOptionalTo204ExceptionMapper());
```
Comment (uschi2000) [inline jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/feignimpl/Java8TestServer.java]: why do we need this?

File: jersey-servers/src/test/resources/test-server.yml [hunk 3]
```diff
@@ -6,6 +6,8 @@ server:
   adminConnectors:
     - type: http
       port: 0
+  gzip:
```
Comment (uschi2000) [inline jersey-servers/src/test/resources/test-server.yml:9]: why this change?

File: jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/feignimpl/Java8TestServer.java (hunk 2 above)
Comment (robert3005) [reply jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/feignimpl/Java8TestServer.java]: Dropwizard 1.x converts Optional.empty to 404 which breaks our tests. While 0.x version did nothing special about Optionals we need this now to preserve the expected behaviour of returning 204

File: jersey-servers/src/test/resources/test-server.yml (hunk 3 above)
Comment (robert3005) [reply jersey-servers/src/test/resources/test-server.yml:9]: Dropwziards gzip filter gets in the way of remoting server compression filter. I think they should do the same thing but dropwizard returns different vary headers

File: jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/feignimpl/GuavaOptionalAwareDecoderTest.java (hunk 1 above)
Comment (robert3005) [reply jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/feignimpl/GuavaOptionalAwareDecoderTest.java]: This is explained in https://github.com/FasterXML/jackson-databind/issues/1264 . Unfortunately previously java.nio.file.Path didn't round trip correctly. In order to fix it they changed the serialization format of Path to always write an absolute path. If you pass a relative path and serialize it will make it absolute which is hard to handle in tests

File: jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/feignimpl/Java8TestServer.java (hunk 2 above)
Comment (robert3005) [reply jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/feignimpl/Java8TestServer.java]: https://github.com/dropwizard/dropwizard/pull/2350 added this class in dropwizard

File: keystores/src/test/java/com/palantir/remoting3/config/ssl/DropwizardSslClientAuthTests.java [hunk 4]
```diff
@@ -52,6 +52,8 @@
                     TestConstants.SERVER_KEY_STORE_JKS_PASSWORD),
             ConfigOverride.config("server.applicationConnectors[0].trustStorePath",
                     TestConstants.CA_TRUST_STORE_PATH.toString()),
+            ConfigOverride.config("server.applicationConnectors[0].trustStorePassword",
```
Comment (robert3005) [inline keystores/src/test/java/com/palantir/remoting3/config/ssl/DropwizardSslClientAuthTests.java:55]: As of https://github.com/eclipse/jetty.project/commit/a8d08df84fb7b6944e3bd11adb724fd05c29e357 truststore password will default to keystore password which will fail upon reading the jks keystore. Internally we require client cert auth truststore to come with passwords

File: jersey-servers/build.gradle [hunk 5]
```diff
@@ -15,4 +15,5 @@ dependencies {
     testCompile "io.dropwizard:dropwizard-testing"
     testCompile "junit:junit"
     testCompile "org.hamcrest:hamcrest-all"
+    testCompile "org.mockito:mockito-core"
```
Comment (robert3005) [inline jersey-servers/build.gradle:18]: You have been relying on transitive mockito-core import from dropwizard that's no longer there

File: jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/feignimpl/GuavaOptionalAwareDecoderTest.java (hunk 1 above)
Comment (uschi2000) [reply jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/feignimpl/GuavaOptionalAwareDecoderTest.java]: ok

File: jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/feignimpl/Java8TestServer.java (hunk 2 above)
Comment (uschi2000) [reply jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/feignimpl/Java8TestServer.java]: we'll need to minimally understand (and maybe deduplicate) this with Java8OptionalMessageBodyWriter and NoContentExceptionMapper

Comment (uschi2000) [reply jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/feignimpl/Java8TestServer.java]: (because it's not obvious how to use a HttpRemotingJerseyFeature.INSTANCE with dropwizard to get the desired behavior)

File: jersey-servers/build.gradle (hunk 5 above)
Comment (uschi2000) [reply jersey-servers/build.gradle:18]: cheers

File: jersey-servers/src/test/resources/test-server.yml (hunk 3 above)
Comment (uschi2000) [reply jersey-servers/src/test/resources/test-server.yml:9]: this is only for CompressionFilterTest, ja?

File: keystores/src/test/java/com/palantir/remoting3/config/ssl/SslConfigurationTest.java [hunk 6]
```diff
@@ -28,18 +28,18 @@
...
                 .trustStoreType(SslConfiguration.StoreType.JKS)
-                .keyStorePath(Paths.get("keystore.jks"))
+                .keyStorePath(Paths.get("/keystore.jks"))
                 .keyStoreType(SslConfiguration.StoreType.JKS)
                 .keyStoreKeyAlias("alias")
                 .keyStorePassword("password")
                 .build();
-        String deserializedCamelCase = "{\"trustStorePath\":\"truststore.jks\",\"trustStoreType\":\"JKS\","
-                + "\"keyStorePath\":\"keystore.jks\",\"keyStorePassword\":\"password\","
+        String deserializedCamelCase = "{\"trustStorePath\":\"file:///truststore.jks\",\"trustStoreType\":\"JKS\","
```
Comment (uschi2000) [inline keystores/src/test/java/com/palantir/remoting3/config/ssl/SslConfigurationTest.java]: can old jackson versions deserialize this? (asking because some services (SMM?) may use the SslConfiguration class over the wire, I believe)

File: keystores/src/test/java/com/palantir/remoting3/config/ssl/DropwizardSslClientAuthTests.java (hunk 4 above)
Comment (uschi2000) [reply keystores/src/test/java/com/palantir/remoting3/config/ssl/DropwizardSslClientAuthTests.java:55]: ok

File: jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/feignimpl/Java8TestServer.java (hunk 2 above)
Comment (uschi2000) [reply jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/feignimpl/Java8TestServer.java]: but then I guess it's called JerseyFeature, not DropwizardFeature. not your fault, let's move on. https://github.com/palantir/http-remoting/issues/714

**Comment by uschi2000** [comment]
looks good modulo the path deserialization wire break question.

File: jersey-servers/src/test/resources/test-server.yml (hunk 3 above)
Comment (robert3005) [reply jersey-servers/src/test/resources/test-server.yml:9]: yes, just for those tests

File: keystores/src/test/java/com/palantir/remoting3/config/ssl/SslConfigurationTest.java (hunk 6 above)
Comment (robert3005) [reply keystores/src/test/java/com/palantir/remoting3/config/ssl/SslConfigurationTest.java]: I think I went overboard here with making behaviour the same in both cases. Let me revert that

Comment (robert3005) [reply keystores/src/test/java/com/palantir/remoting3/config/ssl/SslConfigurationTest.java]: Added extra cases to show the compatibility guarantees

**Review by uschi2000 — APPROVED** [review]

**Comment by uschi2000** [comment]
@iamdanfox do you want to take a look, too, or are you good?

File: extras/jackson-support/src/main/java/com/palantir/remoting3/ext/jackson/ObjectMappers.java [hunk 7]
```diff
@@ -108,7 +108,6 @@ public static ObjectMapper newCborServerObjectMapper() {
     public static ObjectMapper withDefaultModules(ObjectMapper mapper) {
         return mapper
                 .registerModule(new GuavaModule())
-                .registerModule(new ShimJdk7Module())
```
Comment (iamdanfox) [inline extras/jackson-support/src/main/java/com/palantir/remoting3/ext/jackson/ObjectMappers.java]: If this is the only piece of non-test code that was changed then in theory if people hit sad times with the new jackson, they could conceivably force back to a lower version (and re-add this) right?

Comment (j-baker) [reply extras/jackson-support/src/main/java/com/palantir/remoting3/ext/jackson/ObjectMappers.java]: So for this code, there are two arguments:

1. This code makes remoting Jackson version independent. It's why some internal projects run 2.9. So leaving it in is nice because it gives you version independence.
2. OTOH almost everything uses Jackson 2.7+ now, and we might want to use remoting to leverage that.

File: error-handling/versions.lock [hunk 8]
```diff
@@ -1,13 +1,14 @@
 {
     "compileClasspath": {
         "com.fasterxml.jackson.core:jackson-annotations": {
-            "locked": "2.6.7",
+            "locked": "2.9.5",
             "transitive": [
-                "com.fasterxml.jackson.core:jackson-databind"
+                "com.fasterxml.jackson.core:jackson-databind",
+                "com.fasterxml.jackson.datatype:jackson-datatype-jsr310"
```
Comment (uschi2000) [inline error-handling/versions.lock:7]: in 2.8 or 2.9 a lot of this stuff got renamed, no? specifically, why don't we have to update `extras/jackson-support/build.gradle` to reflect the new maven coordinates of the Guava and Java8 data formats?

**Comment by robert3005** [comment]
@iamdanfox @uschi2000 turns out my debugging wasn't quite right - you can override path deserializer and the code we have just works. I mistakenly thought that what was actually a dropwizard misconfiguration was a jackson issue. I have reverted all the path related changes and instead adjusted DW used in test to use `ObjectMappers#newServerObjectMapper` as jackson message body provider

As far as I can tell the source repos moved around but artifacts didn't so we're all good (I just verified all java 8 related packages manually).

File: jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/feignimpl/GuavaOptionalComplexType.java [hunk 9]
```diff
@@ -49,6 +50,15 @@ public Path getPath() {
         return path;
     }
 
+    @Override
+    public String toString() {
```
Comment (robert3005) [inline jaxrs-clients/src/test/java/com/palantir/remoting3/jaxrs/feignimpl/GuavaOptionalComplexType.java:54]: Added for ease of debugging - you get useful assertions errors with this

**Review by uschi2000 — APPROVED** [review]

**Comment by uschi2000** [comment]
thanks, @robert3005
